man/dbugen.1
.TH DBUGEN 1 "12 September 1995"
.FB
.SH NAME
dbugen \- generate updates within a \fIFirstBase\fP database
.SH SYNOPSIS
.B dbugen
[\fB-b\fP] [\fB-y\fP]
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbugen
generates updates within a \fIFirstBase\fP database using an update
dictionary defined via 
\fIdbdupd\fP(1).
This update dictionary is used to determine 
the new value of each of the fields processed.
Fields not mentioned in the update dictionary are not modified.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fI-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
\fIDbugen\fP(1)
first displays an image of the updates that will occur.
The user has a chance to abort here. All updates are done in place, 
directly into the current database.
.PP
\fIDbugen\fP(1)
can change selected fields of selected records on a global basis.
Any field can be cleared out, 
alphanumeric fields can be replaced, and numeric oriented
fields (numeric, float, 
dollar) can actually be formulated during updates.
See 
\fIdbdupd\fP(1)
for more update details.
.SH FILES
.PD 0
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.?dict
dictionary for \fIFirstBase\fP dbase or index.
.TP 10
*.idictu
dictionary for update generator.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.PD
.SH SEE ALSO
dbdupd(1), dbdbas(1), dbedit(1), generators(5).
.PP
.I FirstBase User's Guide and Reference Manual.
.SH DIAGNOSTICS
For those of you who are
more tool oriented, \fIdbdupd\fP(1) and \fIdbugen\fP(1) effectively
implement a macro processor, complete with argument macros, for the
\fIFirstBase\fP.
.sp 1
This tool uses a \fIFirstBase\fP index to process the database in a particular order.
If this index is detected to be out of date or non-existent, it is
regenerated \fIautomatically\fP.
.sp 1
Note: this \fIFirstBase\fP database tool will attempt to gain write access to a record
before operating on that record. If access is currently locked, then the
internal locking routine will respond according to the value of the \fIFirstBase\fP
environment variable LOCKMESSAGE. See \fIsetup\fP(5).
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbugen/dbugen.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbugen.c,v 9.1 2001/02/16 19:48:39 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "%W% %G% FB";
#endif

/*
 *  dbugen.c - update generator. provides mechanism for doing
 *    formula calculations of macros substitution on fields
 *    across an entire fb_database (as defined by an index, of course).
 */

#include <fb.h>
#include <fb_vars.h>
#include <dbd.h>

extern short int cdb_secure;

static fb_upd **itemlist = NULL;	/* for list of items */
static int lastitem = 0;		/* last item in itemlist */
fb_database *hp;

char *T_DELETE = "$DELETE";

static ugen();
static one_ugen();
static update();
static utrace();
static usage();

long n = 0L, atol();

/* 
 *  dbugen - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char fname[FB_MAXNAME];
	 int i;

         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_opendb(cdb_db, READWRITE, FB_ALLINDEX, FB_MUST_INDEX);
         fb_scrhdr(cdb_db, "Parsing"); fb_infoline();
	 sprintf(fname, "%su", hp->idict);
         for (i = 0; i < cdb_db->nfields; i++)
            fb_nounders(cdb_db->kp[i]);
	 itemlist =
	     (fb_upd **) fb_malloc((hp->nfields+1) * sizeof(fb_upd *));
	 			/* allocate enough ptrs for all fields */
	 if ((lastitem = read_idictu(fname, itemlist)) == FB_ERROR)
	    fb_xerror(FB_BAD_DICT, fname, NIL);
	 if (!cdb_batchmode && !cdb_yesmode){
            utrace();
	    if (fb_mustbe('y',"If accurate, enter 'y' to continue: ",
	          cdb_t_lines, 1) == FB_ERROR)
               fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    fb_move(4, 1), fb_clrtobot(), fb_infoline();
	    }
         if (ugen()==FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
         fb_puthead(cdb_db->fd, cdb_db->reccnt, cdb_db->delcnt);
	 fb_closedb(cdb_db);
         fb_ender();
      }
      
/* 
 *  ugen - generate converted data base  according to list in pp 
 */
 
   static ugen()
      
      {

         int max;

         if (!cdb_batchmode){
            fb_scrstat("Updating");
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
         fb_forxeach(hp, one_ugen);
         return(FB_AOK);
      }

   static one_ugen(hp)
      fb_database *hp;

      {
         long rec;

         fb_set_autoindex(hp);
         rec = hp->rec;
         ++n;
         if (!cdb_batchmode)
            fb_gcounter(n);
         if (fb_lock(rec, hp, FB_NOWAIT) != FB_AOK) /* NOWAIT if locked */
            return(FB_AOK);
         update();
         fb_lock_head(hp);
         fb_setdirty(hp, 1);
         if (fb_putrec(rec, hp) == FB_ERROR)
            fb_xerror(FB_FATAL_PUTREC, hp->dbase, (char *) &rec);
         fb_setdirty(hp, 0);
	 if (fb_put_autoindex(hp) == FB_ERROR)
	    fb_serror(FB_BAD_INDEX, hp->dbase, NIL);
	 fb_allsync(hp);
         fb_unlock_head(hp);
         fb_unlock(rec, hp);
         return(FB_AOK);
      }

/* 
 *  update - update the buffer according to the itemlist of upd structs 
 */
 
   static update()

      {
         fb_upd *up;
	 int i;
	 
	 for (i = 1; i <= lastitem; i++){
	    up = itemlist[i];
	    cdb_bfld[0] = NULL;
	    fb_trim(up->fv);		/* not needed i think ??? */
	    if (up->fv[0] == NULL)
	       ;
	    else if (equal(up->fv, T_DELETE)){
               cdb_db->kp[cdb_db->nfields]->fld[0] = CHAR_STAR;
	       cdb_db->delcnt++;
	       continue;
	       }
	    else if (!(FB_OFNUMERIC(up->fp->type))){
	       if (up->fv != NULL)
	          strcpy(cdb_bfld, up->fv);
	       }
	    else{			/* numeric type can be formulated */
	       if (fb_getformula(up->fp, up->fv, cdb_bfld, 0, hp) == FB_ERROR)
	          fb_xerror(FB_BAD_DATA, SYSMSG[S_BAD_FORMULA], NIL);
	       }
	    fb_store(up->fp, cdb_bfld, hp);
	    }
      }

/*  
 *  utrace - trace an itemlist of upd structures to the screen 
 */
 
   static utrace()
      {
      
	 int i;
	 fb_upd *up;
	 int row;
	    
	 for (i = 1; i <= lastitem; ){
	    fb_move(4, 1);
	    for (row = 4; i <= lastitem && row <= 22; row += 2){
	       up = itemlist[i++];
	       fb_printw("%10s %c %6d ---> ", up->fp->id, up->fp->type, 
		     up->fp->size);
	       if (up->fv != NULL)
		  fb_printw(FB_FSTRING, up->fv);
	       else
	          fb_printw("<EMPTY>");
	       fb_printw("\n");
	       }
	    }
      }
			
/* 
 *  usage message 
 */
 
   static usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbugen [-d dbase] [-i index] [-b] [-y]",
            NIL);
      }
